### PR TITLE
[SPARK-11097][Core]Add channelActive callback to RpcHandler to monitor the new connections

### PR DESCRIPTION
**Pull request by zsxwing**
Added `channelActive` to `RpcHandler` so that `NettyRpcHandler` doesn't need `clients` any more.

### CODE REVIEW

**Comment by zsxwing** [comment]
@vanzin could you take a look once you have time? I added three tests in this PR to make sure the network events can be reported in different cases.

**Comment by vanzin** [comment]
It's unlikely I'll have time to review much of anything for the next couple of weeks... sorry.

File: network/common/src/main/java/org/apache/spark/network/server/MessageHandler.java [hunk 1]
```diff
@@ -28,6 +28,9 @@
   /** Handles the receipt of a single message. */
   public abstract void handle(T message) throws Exception;
 
+  /** Invoked when the channel this MessageHandler is on is active. */
+  public abstract void channelActive();
```
Comment (rxin) [inline network/common/src/main/java/org/apache/spark/network/server/MessageHandler.java]: is this a different event from RpcHandler's connectionEstablished? If they are the same one, can we name them the same?

File: network/common/src/main/java/org/apache/spark/network/server/TransportChannelHandler.java [hunk 2]
```diff
@@ -84,6 +84,22 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws E
...
+      logger.error("Exception from request handler while registering channel", e);
+    }
+    try {
+      responseHandler.channelActive();
+    } catch (RuntimeException e) {
+      logger.error("Exception from response handler while registering channel", e);
+    }
+    super.channelRegistered(ctx);
+  }
+
```
Comment (rxin) [inline network/common/src/main/java/org/apache/spark/network/server/TransportChannelHandler.java]: remove the extra blank line here

File: network/common/src/main/java/org/apache/spark/network/server/MessageHandler.java (hunk 1 above)
Comment (rxin) [inline network/common/src/main/java/org/apache/spark/network/server/MessageHandler.java]: also channelActive vs channelUnregistered is somewhat confusing. Either channelActive vs channelInactive, or channelRegistered vs channelUnregistered, or channelConnected vs channelDisconnected.

**Comment by rxin** [comment]
@andrewor14 can you take a look also?

**Comment by zsxwing** [comment]
I just renamed these methods to channelActive & channelInactive to be consistent with Netty.

File: network/common/src/main/java/org/apache/spark/network/server/TransportChannelHandler.java (hunk 2 above)
Comment (zsxwing) [inline network/common/src/main/java/org/apache/spark/network/server/TransportChannelHandler.java]: > remove the extra blank line here

done

**Comment by andrewor14** [comment]
LGTM, merging into master.

**Pull request closed without merging**